pkg/profiler/cpu/cpu.go

```diff
@@ -549,40 +566,47 @@ func (p *CPU) Run(ctx context.Context) error {
 		p.metrics.obtainDuration.Observe(time.Since(obtainStart).Seconds())
 
 		processLastErrors := map[int]error{}
-		for _, prof := range profiles {
-			processLastErrors[int(prof.ID.PID)] = nil
+		for _, perProcessRawData := range rawData {
+			pid := int(perProcessRawData.PID)
+			processLastErrors[pid] = nil
 
-			if err := p.symbolizer.Symbolize(prof); err != nil {
-				// This could be a partial symbolization, so we still want to send the profile.
-				level.Debug(p.logger).Log("msg", "failed to symbolize profile", "pid", prof.ID.PID, "err", err)
-				processLastErrors[int(prof.ID.PID)] = err
-			}
-
-			// ConvertToPprof can combine multiple profiles into a single profile,
-			// however right now we chose to send each profile separately.
-			// This is not too inefficient as we batch the profiles in a single RPC message,
-			// using the batch profiler writer.
-			pprof, err := profiler.ConvertToPprof(p.LastProfileStartedAt(), samplingPeriod, prof)
+			pi, err := p.processInfoManager.Info(ctx, pid)
 			if err != nil {
-				level.Warn(p.logger).Log("msg", "failed to convert profile to pprof", "pid", prof.ID.PID, "err", err)
-				processLastErrors[int(prof.ID.PID)] = err
+				p.metrics.profileDrop.WithLabelValues(profileDropReasonProcessInfo).Inc()
+				level.Warn(p.logger).Log("msg", "failed to get process info", "pid", pid, "err", err)
+				processLastErrors[pid] = err
 				continue
 			}
 
-			pi, err := p.processInfoManager.Info(ctx, int(prof.ID.PID))
+			pprof, err := pprof.NewConverter(
```

Comment:
Why don't we inject the converter as a dependency rather than injecting its dependencies?

We can have a generic interface for converters to pass to the `ProfileWriter`
https://github.com/parca-dev/parca-agent/blob/ae203681363ffc2690a0da2c65a94b7db8883b38/pkg/profiler/profiler.go#L133-L137

Gets `RawSamples` and produces a `[]byte` or `io.Writer`. It would be cleaner, and we could change the wire format. I think this is what you had in mind.

Reply:
Yeah I agree, anything beyond raw samples should not live in the CPU profiler. Up to you if you want me to do that here or in a follow up.

Reply:
Follow-ups are totally fine!